fmpz/aors_ui.c

```diff
@@ -0,0 +1,338 @@
+                    _fmpz_clear_mpz(f1);
+                *f = g1;
+                return;
+            }
+nocarry:    sz = 1; /* No carry, but result does not fit in small fmpz */
+carry:      if (COEFF_IS_MPZ(f1))
+                mf = COEFF_TO_PTR(f1);
+            else
+            {
+                mf = _fmpz_new_mpz();
+                *f = PTR_TO_COEFF(mf);
+            }
+            mf->_mp_size = sz;
+            mf->_mp_d[0] = g1;
+            mf->_mp_d[1] = 1; /* Set carry (not used if sz = 1) */
+        }
+        else /* g < 0 */
+        {
+            g1 += x;
+            if (((slong) x) >= 0 && g1 <= COEFF_MAX)
+            {
+                /* If x > 0 does not have its top bit set
+                 * and COEFF_MIN <= g < 0, we can interpret x + g as a slong.
+                 * So if the result in g1 is smaller than COEFF_MAX, it is a
+                 * small fmpz. */
+                if (COEFF_IS_MPZ(f1))
+                    _fmpz_clear_mpz(f1);
+                *f = g1;
+                return;
+            }
+            else
+            {
+                /* 1) If top bit is set in x, the result is going to be positive
+                 *    but will be larger than COEFF_MAX since
+                 *
+                 *          x + g  >=  (2^63) - (2^62 - 1)  =  2^62 + 1.
+                 *
+                 *    However, it will be contained in one limb since g < 0.
+                 *
+                 * 2) If top bit is not set, then result is larger than
+                 *    COEFF_MAX, and so it cannot be a small fmpz. However, it
+                 *    must fit in one limb since
+                 *
+                 *          x + g  <=  (2^63 - 1) + (-1)  =  2^63 - 2,
+                 *
+                 *    which is contained in one limb. */
+
+                goto nocarry;
+            }
+        }
+    }
+    else
+    {
+        __mpz_struct * mg = COEFF_TO_PTR(g1);
+        mp_size_t gsz = mg->_mp_size;
+        mp_limb_t * glimbs = mg->_mp_d;
+
+        if (gsz > 0)
+            _fmpz_add_mpn_1(f, glimbs, gsz, x);
+        else
+            _fmpz_sub_mpn_1(f, glimbs, gsz, x);
+    }
+}
+
+/* "Add" two number with same sign. Decide sign from g. */
+static void
+_fmpz_add_mpn_1(fmpz_t f, const mp_limb_t * glimbs, mp_size_t gsz, mp_limb_t x)
+{
+    __mpz_struct * mf;
+    mp_limb_t * flimbs;
+    mp_size_t gabssz = FLINT_ABS(gsz);
+
+    /* Promote f as it is guaranteed to be large */
+    if (COEFF_IS_MPZ(*f))
+        mf = COEFF_TO_PTR(*f);
+    else
+    {
+        mf = _fmpz_new_mpz();
+        *f = PTR_TO_COEFF(mf);
+    }
+    flimbs = mf->_mp_d;
+
+    if (mf->_mp_alloc < (gabssz + 1)) /* Ensure result fits */
+    {
+        mf->_mp_d = realloc(mf->_mp_d, sizeof(mp_limb_t) * (gabssz + 1));
+        mf->_mp_alloc = gabssz + 1;
+
+        /* If f and g are aliased, then we need to change glimbs as well. */
+        if (flimbs == glimbs)
+            glimbs = mf->_mp_d;
```

Comment:
Somehow this does not get tested in the tests for these functions (even when number of tests for aliasing is huge), but get tested in `divides_mod_list`, for which it passes.